### PR TITLE
docs: add comment about file changes

### PR DESCRIPTION
**Pull request by daixiang0**
Signed-off-by: Loong Dai <loong.dai@intel.com>

Commit Message:
Additional Description:
Risk Level: Low
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA] #20474
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: docs/root/start/quick-start/configuration-dynamic-filesystem.rst [hunk 1]
```diff
@@ -8,6 +8,11 @@ protocol.
 
 When the files are changed on the filesystem, Envoy will automatically update its configuration.
 
+Notice:
+
+Now Enovy only get the changes when write to temporary file and rename into place, ignore overwritten directly.
+The purpose is  that atomic renames are the only way to achieve consistency.
+
```
Comment (wbpcode) [inline docs/root/start/quick-start/configuration-dynamic-filesystem.rst:15]: Thanks. May be a note `  .. note::` would be better.

**Review by wbpcode — COMMENTED** [review]
Thanks. Nit comment added.

File: docs/root/start/quick-start/configuration-dynamic-filesystem.rst (hunk 1 above)
Comment (daixiang0) [reply docs/root/start/quick-start/configuration-dynamic-filesystem.rst:15]: Sure, updated.

**Comment by phlax** [comment]
@daixiang0 is this correct ? Im wondering because the dynamic fs sandbox edits in place (by way of `sed -i`) and seems to still be working ok

**Comment by phlax** [comment]
(my suspicion is that this is more some kind of mac/inode issue)

**Comment by phlax** [comment]
ok - i see the related issue here https://github.com/envoyproxy/envoy/issues/20474

i guess this works with `sed` as it must do some kind of tempwrite/replace

File: docs/root/start/quick-start/configuration-dynamic-filesystem.rst [hunk 2]
```diff
@@ -8,6 +8,11 @@ protocol.
 
 When the files are changed on the filesystem, Envoy will automatically update its configuration.
 
+.. note::
+
+   Now Enovy only get the changes when write to temporary file and rename into place, ignore overwritten directly.
+   The purpose is  that atomic renames are the only way to achieve consistency.
```
Comment (phlax) [inline docs/root/start/quick-start/configuration-dynamic-filesystem.rst:14]: The grammar is not quite correct here and (esp for a beginner doc) this is not so easy to follow

Suggested change:

> Envoy only updates when the configuration file is replaced, and not when the file is edited in place. It is implemented this way to ensure configuration consistency.

**Review by phlax — CHANGES_REQUESTED** [review]
@daixiang0 thanks for adding this - would be good to simplify/clarify the note a little

File: docs/root/start/quick-start/configuration-dynamic-filesystem.rst (hunk 2 above)
Comment (mattklein123) [reply docs/root/start/quick-start/configuration-dynamic-filesystem.rst:14]: Please also ref link to this section of the docs, which is a general explanation of how the atomic swap works: https://www.envoyproxy.io/docs/envoy/latest/configuration/operations/runtime#updating-runtime-values-via-symbolic-link-swap

Comment (daixiang0) [reply docs/root/start/quick-start/configuration-dynamic-filesystem.rst]: @phlax and @mattklein123, thanks for the review. I have updated.

File: docs/root/start/quick-start/configuration-dynamic-filesystem.rst [hunk 3]
```diff
@@ -90,3 +95,8 @@ proxies over ``TLS`` to https://www.envoyproxy.io.
     :language: yaml
     :linenos:
     :emphasize-lines: 11, 17-18, 22-23
+
+.. seealso::
+
+   :ref:`Atomic swap <_config_runtime_symbolic_link_swap>`
```
Comment (phlax) [inline docs/root/start/quick-start/configuration-dynamic-filesystem.rst:101]: im not sure if this is immediately clear as to the relevance

perhaps we can change the link title to something like

> Atomic swap for configuration file updates

**Review by phlax — CHANGES_REQUESTED** [review]

File: docs/root/start/quick-start/configuration-dynamic-filesystem.rst (hunk 3 above)
Comment (mattklein123) [reply docs/root/start/quick-start/configuration-dynamic-filesystem.rst:101]: I still think this should just be ref linked above. It seems fine to do inline.

/wait

**Comment by daixiang0** [comment]
There it is:

```
docs/root/configuration/operations/runtime.rst:127:.. _config_runtime_symbolic_link_swap:
```

BTW, how can I get right labels?

File: docs/root/start/quick-start/configuration-dynamic-filesystem.rst [hunk 4]
```diff
@@ -8,6 +8,15 @@ protocol.
 
 When the files are changed on the filesystem, Envoy will automatically update its configuration.
 
+.. seealso::
+
+   :ref:`Atomic swap <_config_runtime_symbolic_link_swap>`
```
Comment (wbpcode) [inline docs/root/start/quick-start/configuration-dynamic-filesystem.rst:13]: Please remove the prefix `_` in the link.

Comment (daixiang0) [reply docs/root/start/quick-start/configuration-dynamic-filesystem.rst]: I will try.

Comment (daixiang0) [reply docs/root/start/quick-start/configuration-dynamic-filesystem.rst]: @wbpcode thanks, it helps.

File: docs/root/start/quick-start/configuration-dynamic-filesystem.rst [hunk 5]
```diff
@@ -8,6 +8,11 @@ protocol.
 
 When the files are changed on the filesystem, Envoy will automatically update its configuration.
 
+.. note::
+
+   Envoy only updates when the configuration file is replaced, and not when the file is edited in place.
+   It is implemented this way to ensure configuration consistency, more details refer to :ref:`Atomic swap for configuration file updates<config_runtime_symbolic_link_swap>`.
```
Comment (phlax) [inline docs/root/start/quick-start/configuration-dynamic-filesystem.rst:14]: The grammar is not quite right here, and i think it needs some spacing

```suggestion
   Envoy only updates when the configuration file is replaced, and not when the file is edited in place.
   
   It is implemented this way to ensure configuration consistency.
   
   Please see the section on :ref:`atomic swaps <config_runtime_symbolic_link_swap>` for further information on how runtime configuration is updated.
```

File: docs/root/start/quick-start/configuration-dynamic-filesystem.rst [hunk 6]
```diff
@@ -90,3 +95,4 @@ proxies over ``TLS`` to https://www.envoyproxy.io.
     :language: yaml
     :linenos:
     :emphasize-lines: 11, 17-18, 22-23
+
```
Comment (phlax) [inline docs/root/start/quick-start/configuration-dynamic-filesystem.rst:98]: nit: unnecessary ws

**Review by phlax — CHANGES_REQUESTED** [review]

File: docs/root/start/quick-start/configuration-dynamic-filesystem.rst (hunk 3 above)
Comment (phlax) [reply docs/root/start/quick-start/configuration-dynamic-filesystem.rst]: @mattklein123 im also a bit concerned that this link is confusing (esp to a beginner) as when i click through its related to the runtime fs and there isnt any information there about updating configuration files

**Comment by phlax** [comment]
testing this further - i think we probably need to update the docs to make clear that only an fs `mv` will work - afaict `cp` also will not work

eg, doing the following works

```console
root@00eb0e6c500b:~# cp /var/lib/envoy/cds.yaml .
root@00eb0e6c500b:~# sed -i s/service1/service2/ cds.yaml
root@00eb0e6c500b:~# mv cds.yaml /var/lib/envoy/
```

whereas this does not:

```console
root@00eb0e6c500b:~# cp /var/lib/envoy/cds.yaml .
root@00eb0e6c500b:~# sed -i s/service1/service2/ cds.yaml
root@00eb0e6c500b:~# cp cds.yaml /var/lib/envoy/
```

**Comment by phlax** [comment]
also - as it stands the note ~contradicts what the docs say

the section header

> Step 4: Edit cds.yaml inside the container...

whereas the note says...

> ...and not when the file is edited in place.

i guess we can change the section title to "Step 4: _Replace_ cds.yaml..." - but then i also think we need to give at least a brief explanation as to why `sed -i` (ie inplace) works

File: docs/root/start/quick-start/configuration-dynamic-filesystem.rst (hunk 3 above)
Comment (mattklein123) [reply docs/root/start/quick-start/configuration-dynamic-filesystem.rst]: OK whatever you think this is best is fine with me. I was thinking something along the lines of: "Envoy will only watch configuration files for atomic moves. An example of how to set this up is described [here]."

Comment (phlax) [reply docs/root/start/quick-start/configuration-dynamic-filesystem.rst]: my suggestion would be:

- add a bit more info wrt cp/mv and why the sed example works (re https://github.com/envoyproxy/envoy/pull/20475#issuecomment-1079019891 and following comment)
- move the link to see also at the bottom

File: docs/root/start/quick-start/configuration-dynamic-filesystem.rst (hunk 6 above)
Comment (daixiang0) [reply docs/root/start/quick-start/configuration-dynamic-filesystem.rst:98]: Done.

File: docs/root/start/quick-start/configuration-dynamic-filesystem.rst (hunk 5 above)
Comment (daixiang0) [reply docs/root/start/quick-start/configuration-dynamic-filesystem.rst:14]: Done.

**Comment by daixiang0** [comment]
"sed -i" means replace in place, not edit in place, like the comment in the commits.

I do not know more details, maybe matt knows.

/cc @mattklein123

**Comment by phlax** [comment]
> means replace in place, not edit in place

`sed -i` == "stream editor, inplace"

**Comment by daixiang0** [comment]
Check the man page, it does say edit in place.

**Comment by daixiang0** [comment]
@mattklein123 thanks for your review.

**Comment by daixiang0** [comment]
ping @phlax

File: docs/root/start/quick-start/configuration-dynamic-filesystem.rst [hunk 7]
```diff
@@ -8,6 +8,14 @@ protocol.
 
 When the files are changed on the filesystem, Envoy will automatically update its configuration.
 
+.. note::
+
+   Envoy only updates when the configuration file is replaced, and not when the file is edited in place.
```
Comment (phlax) [inline docs/root/start/quick-start/configuration-dynamic-filesystem.rst:13]:
```suggestion
   Envoy only updates when the configuration file is replaced by a file move, and not when the file is edited in place.
```

File: docs/root/start/quick-start/configuration-dynamic-filesystem.rst [hunk 8]
```diff
@@ -8,6 +8,14 @@ protocol.
...
+
+   It is implemented this way to ensure configuration consistency.
+
+   Please see the section on :ref:`atomic swaps <config_runtime_symbolic_link_swap>` for further information on how runtime configuration is updated.
```
Comment (phlax) [inline docs/root/start/quick-start/configuration-dynamic-filesystem.rst:17]: lets move this back to the the sealso section

File: docs/root/start/sandboxes/dynamic-configuration-filesystem.rst [hunk 9]
```diff
@@ -100,6 +100,13 @@ You can do this using ``sed`` inside the container:
 
    docker-compose exec -T proxy sed -i s/service1/service2/ /var/lib/envoy/cds.yaml
 
+Notice that `sed -i` means replace not edit in place.
+You can create a new file and useing `mv` to do:
+
+.. code-block:: console
```
Comment (phlax) [inline docs/root/start/sandboxes/dynamic-configuration-filesystem.rst:106]: this is unnecessary - and incorrect - if we were to add it we would need to also add to verify.sh - but rather just remove

File: docs/root/start/sandboxes/dynamic-configuration-filesystem.rst [hunk 10]
```diff
@@ -100,6 +100,13 @@ You can do this using ``sed`` inside the container:
 
    docker-compose exec -T proxy sed -i s/service1/service2/ /var/lib/envoy/cds.yaml
 
+Notice that `sed -i` means replace not edit in place.
+You can create a new file and useing `mv` to do:
```
Comment (phlax) [inline docs/root/start/sandboxes/dynamic-configuration-filesystem.rst:104]:
```suggestion
Notice that `sed -i` means replace not edit in place.
You can create a new file and useing `mv` to do:
```

```
.. note::

    The above example uses ``sed -i``, which works as an inplace edit as `sed` does copy, edit and move in order to do this.

```

**Review by phlax — CHANGES_REQUESTED** [review]

File: docs/root/start/quick-start/configuration-dynamic-filesystem.rst (hunk 7 above)
Comment (daixiang0) [reply docs/root/start/quick-start/configuration-dynamic-filesystem.rst:13]: Done.

File: docs/root/start/quick-start/configuration-dynamic-filesystem.rst (hunk 8 above)
Comment (daixiang0) [reply docs/root/start/quick-start/configuration-dynamic-filesystem.rst:17]: Done.

File: docs/root/start/sandboxes/dynamic-configuration-filesystem.rst (hunk 9 above)
Comment (daixiang0) [reply docs/root/start/sandboxes/dynamic-configuration-filesystem.rst:106]: Removed.

File: docs/root/start/sandboxes/dynamic-configuration-filesystem.rst (hunk 10 above)
Comment (daixiang0) [reply docs/root/start/sandboxes/dynamic-configuration-filesystem.rst:104]: Done.

**Comment by daixiang0** [comment]
> Code coverage for source/common/http is lower than limit of 96.3 (96.2)

Seems it is introduced by other PRs by force merge, could we ignore it?

**Review by phlax — APPROVED** [review]
lgtm, thanks @daixiang0 for addressing this

**Pull request merged**